### PR TITLE
Add the method return type and the parameter type in the mock generator.

### PR DESCRIPTION
**Pull request by GuillaumeDievart**

### CODE REVIEW

File: classes/mock/generator.php [hunk 1]
```diff
@@ -805,77 +818,85 @@ protected static function generateUnknownClassCode($class, $mockNamespace, $mock
...
+				'static',
+				'switch',
+				'throw',
+				'trait',
+				'try',
+				'unset',
+				'use',
+				'var',
+				'while',
+				'xor',
```
Comment (Hywan) [inline classes/mock/generator.php]: What's new here?

File: classes/mock/generator.php [hunk 2]
```diff
@@ -805,77 +818,85 @@ protected static function generateUnknownClassCode($class, $mockNamespace, $mock
...
+			);
 		}
+
+		$methodNameReservedWord[] =	 '__halt_compiler';
```
Comment (Hywan) [inline classes/mock/generator.php]: Remove extra space after `=` please.

File: classes/mock/generator.php [hunk 3]
```diff
@@ -805,77 +818,85 @@ protected static function generateUnknownClassCode($class, $mockNamespace, $mock
...
+
+		return $methodNameReservedWord;
+
```
Comment (Hywan) [inline classes/mock/generator.php]: Remove this empty line plese.

File: classes/mock/generator.php [hunk 4]
```diff
@@ -679,6 +689,9 @@ protected static function getParameterType(\reflectionParameter $parameter)
 			case ($class = $parameter->getClass()):
 				return '\\' . $class->getName() . ' ';
 
+			case method_exists($parameter, 'hasType') && $parameter->hasType():
+				return (string)$parameter->getType() . ' ';
```
Comment (Hywan) [inline classes/mock/generator.php]: I am not sure if we need an extra space after a cast with atoum's conventions.

**Comment by Hywan** [comment]
Interesting work! Thanks. Few CS and precisions are needed though.

File: classes/mock/generator.php (hunk 4 above)
Comment (GuillaumeDievart) [inline classes/mock/generator.php]: The cast is useless with the concatenation, I will remove it.

File: classes/mock/generator.php (hunk 1 above)
Comment (GuillaumeDievart) [inline classes/mock/generator.php]: Nothing of new. But these keywords are now allowed to use in the methods name: http://3v4l.org/44nJi

So, I exclude them from the reversedWord in PHP7. And yield and Finally keywords are missing for PHP5.5.

Comment (Hywan) [inline classes/mock/generator.php]: Ok so you added `yield` and `finally`?

File: classes/mock/generator.php (hunk 4 above)
Comment (Hywan) [inline classes/mock/generator.php]: Indeed!

File: classes/mock/generator.php (hunk 1 above)
Comment (GuillaumeDievart) [inline classes/mock/generator.php]: I don't think this modification must be include in this PR. 

These keywords (yield and finally) are disallowed in 
@php >= 5.5 
@php < 7.0

File: classes/mock/generator.php [hunk 5]
```diff
@@ -805,77 +818,84 @@ protected static function generateUnknownClassCode($class, $mockNamespace, $mock
...
-			default:
-				return false;
+		return in_array($method->getName(), self::getMethodNameReservedWordByVersion(), true);
+	}
+
+	protected static function getMethodNameReservedWordByVersion()
+	{
+		$methodNameReservedWord = array();
+
+		if(PHP_MAJOR_VERSION < 7) {
```
Comment (jubianchi) [inline classes/mock/generator.php]: For readability, I would have reversed this ìf` : 

```php
if(PHP_MAJOR_VERSION >= 7) {
    return array('__halt_compiler');
}

return array(
    // All the reserved words for php < 7
);
```

File: tests/units/classes/mock/generator.php [hunk 6]
```diff
@@ -1680,6 +2335,24 @@ public function testMethodIsMockableWithReservedWord($reservedWord)
...
+			->and($this->mockGenerator->orphanize('__construct'))
+			->and($method = new \mock\reflectionMethod($this, $reservedWord))
+			->and($this->calling($method)->getName = $reservedWord)
+			->and($this->calling($method)->isFinal = false)
+			->and($this->calling($method)->isStatic = false)
+			->and($this->calling($method)->isAbstract = false)
+			->and($this->calling($method)->isPrivate = false)
+			->and($this->calling($method)->isProtected = false)
+			->then
+			->boolean($generator->methodIsMockable($method))->isFalse()
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an identation level

File: tests/units/classes/mock/generator.php [hunk 7]
```diff
@@ -1461,6 +2037,84 @@ public function testGetMockedClassCodeForRealClassWithoutConstructor()
...
+			->and($reflectionClassController->getMethods = array($reflectionMethod))
+			->and($reflectionClassController->getConstructor = null)
+			->and($reflectionClassController->isAbstract = false)
+			->and($reflectionClass = new \mock\reflectionClass(null))
+			->and($generator->setReflectionClassFactory(function() use ($reflectionClass) { return $reflectionClass; }))
+			->and($adapter = new atoum\test\adapter())
+			->and($adapter->class_exists = function($class) use (& $realClass) { return ($class == '\\' . $realClass); })
+			->and($generator->setAdapter($adapter))
+			->then
+			->string($generator->getMockedClassCode($realClass = uniqid()))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 8]
```diff
@@ -2419,6 +3104,254 @@ public function testGetMockedClassCodeForClassWithOnlyVariadicArgumentsInMethod(
...
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 9]
```diff
@@ -2419,6 +3104,254 @@ public function testGetMockedClassCodeForClassWithOnlyVariadicArgumentsInMethod(
...
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 10]
```diff
@@ -2419,6 +3104,254 @@ public function testGetMockedClassCodeForClassWithOnlyVariadicArgumentsInMethod(
...
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 11]
```diff
@@ -1385,6 +1790,177 @@ public function testGetMockedClassCodeForInterfaceWithStaticMethod()
...
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 12]
```diff
@@ -1385,6 +1790,177 @@ public function testGetMockedClassCodeForInterfaceWithStaticMethod()
...
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 13]
```diff
@@ -1093,6 +1314,187 @@ public function testGetMockedClassCodeForInterface()
...
+			->and($reflectionClassController->isInstantiable = false)
+			->and($reflectionClassController->implementsInterface = false)
+			->and($reflectionClass = new \mock\reflectionClass(null))
+			->and($generator->setReflectionClassFactory(function() use ($reflectionClass) { return $reflectionClass; }))
+			->and($adapter = new atoum\test\adapter())
+			->and($adapter->class_exists = function($class) use (& $realClass) { return ($class == '\\' . $realClass); })
+			->and($generator->setAdapter($adapter))
+			->and($generator->disallowUndefinedMethodUsage())
+			->then
+			->string($generator->getMockedClassCode($realClass = uniqid()))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 14]
```diff
@@ -1093,6 +1314,187 @@ public function testGetMockedClassCodeForInterface()
...
+				"\t" . '{' . PHP_EOL .
+				"\t\t" . 'return ' . var_export(array('__construct', '__call'), true) . ';' . PHP_EOL .
+				"\t" . '}' . PHP_EOL .
+				'}' . PHP_EOL .
+				'}'
+			)
+			->if($reflectionClassController->implementsInterface = function($interface) { return ($interface == 'traversable' ? true : false); })
+			->and($generator->setReflectionClassFactory(function($class) use ($reflectionClass) { return ($class == 'iteratorAggregate' ? new \reflectionClass('iteratorAggregate') : $reflectionClass); }))
+			->then
+			->string($generator->getMockedClassCode($realClass = uniqid()))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 15]
```diff
@@ -1093,6 +1314,187 @@ public function testGetMockedClassCodeForInterface()
...
+			->and($reflectionClassController->getMethods = array($reflectionMethod))
+			->and($reflectionClassController->isInstantiable = false)
+			->and($reflectionClassController->implementsInterface = false)
+			->and($reflectionClass = new \mock\reflectionClass(null))
+			->and($generator->setReflectionClassFactory(function() use ($reflectionClass) { return $reflectionClass; }))
+			->and($adapter = new atoum\test\adapter())
+			->and($adapter->class_exists = function($class) use (& $realClass) { return ($class == '\\' . $realClass); })
+			->and($generator->setAdapter($adapter))
+			->then
+			->string($generator->getMockedClassCode($realClass = uniqid()))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 16]
```diff
@@ -788,6 +886,128 @@ public function testGetMockedClassCodeWithAllIsInterface()
...
+				"\t" . '{' . PHP_EOL .
+				"\t\t" . 'return ' . var_export(array('__construct', 'foo'), true) . ';' . PHP_EOL .
+				"\t" . '}' . PHP_EOL .
+				'}' . PHP_EOL .
+				'}'
+			)
+
+			->if($generator->testedClassIs($realClass))
+			->then
+			->string($generator->getMockedClassCode($realClass))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 17]
```diff
@@ -788,6 +886,128 @@ public function testGetMockedClassCodeWithAllIsInterface()
...
+			->and($reflectionClassController->isAbstract = false)
+			->and($reflectionClass = new \mock\reflectionClass(null))
+			->and($generator->setReflectionClassFactory(function() use ($reflectionClass) { return $reflectionClass; }))
+			->and($adapter = new atoum\test\adapter())
+			->and($adapter->class_exists = function($class) use ($realClass) { return ($class == '\\' . $realClass); })
+			->and($generator->setAdapter($adapter))
+			->and($generator->shunt('__construct'))
+			->and($generator->allIsInterface())
+			->then
+			->string($generator->getMockedClassCode($realClass))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

File: tests/units/classes/mock/generator.php [hunk 18]
```diff
@@ -470,6 +471,102 @@ public function testGetMockedClassCodeForRealClassWithCallsToParentClassShunted(
...
+			->and($reflectionClassController->getMethods = array($reflectionMethod, $otherReflectionMethod))
+			->and($reflectionClassController->getConstructor = $reflectionMethod)
+			->and($reflectionClass = new \mock\reflectionClass(null))
+			->and($generator->setReflectionClassFactory(function() use ($reflectionClass) { return $reflectionClass; }))
+			->and($adapter = new atoum\test\adapter())
+			->and($adapter->class_exists = function($class) use (& $realClass) { return ($class == '\\' . $realClass); })
+			->and($generator->setAdapter($adapter))
+			->and($generator->shuntParentClassCalls())
+			->then
+			->string($generator->getMockedClassCode($realClass = uniqid()))->isEqualTo(
```
Comment (jubianchi) [inline tests/units/classes/mock/generator.php]: missing an indentation level

**Comment by jubianchi** [comment]
@GuillaumeDievart minor CS issues and missing changelog entry. Everything else seems good to me. 

:+1:  thanks for the work!

**Comment by jubianchi** [comment]
@GuillaumeDievart awesome!

I'll wait until tomorrow before merging if anybody has something to say it's time to do so :)

**Comment by GuillaumeDievart** [comment]
OK !

File: classes/mock/generator.php [hunk 19]
```diff
@@ -602,6 +602,16 @@ protected function getNamespace($class)
 		return '\\' . $this->getDefaultNamespace() . ($lastAntiSlash === false ? '' : '\\' . substr($class, 0, $lastAntiSlash));
 	}
 
+	protected function getReturnType(\reflectionMethod $method)
+	{
+		$returnTypeCode = '';
```
Comment (jubianchi) [inline classes/mock/generator.php]: please, could you add a blank line between this line and the `if` ?

File: classes/mock/generator.php [hunk 20]
```diff
@@ -602,6 +602,16 @@ protected function getNamespace($class)
...
+		if($method->getName() !== '__construct' && method_exists($method, 'hasReturnType') && $method->hasReturnType()) {
```
Comment (jubianchi) [inline classes/mock/generator.php]: the opening `{` should be on its own line

File: classes/mock/generator.php [hunk 21]
```diff
@@ -805,77 +818,81 @@ protected static function generateUnknownClassCode($class, $mockNamespace, $mock
...
-			case 'xor':
-				return true;
+		return in_array($method->getName(), self::getMethodNameReservedWordByVersion(), true);
+	}
 
-			default:
-				return false;
+	protected static function getMethodNameReservedWordByVersion()
+	{
+		if(PHP_MAJOR_VERSION >= 7) {
```
Comment (jubianchi) [inline classes/mock/generator.php]: the opening `{` should be on its own line

**Comment by jubianchi** [comment]
@GuillaumeDievart sorry I just found some new CS issues. Once they get fixed, I'll merge the PR.

Sorry and thanks again.

**Comment by GuillaumeDievart** [comment]
No problem @jubianchi, it's my fault. I should have taken more time to read http://docs.atoum.org/fr/latest/participer.html#convention-de-codage

File: CHANGELOG.md [hunk 22]
```diff
@@ -92,3 +93,4 @@
 [@jubianchi]: https://github.com/jubianchi
 [@hywan]: https://github.com/hywan
 [@metalaka]: https://github.com/metalaka
+[@GuillaumeDievart]: https://github.com/GuillaumeDievart
```
Comment (Hywan) [inline CHANGELOG.md]: The case does not match `@G…` vs. `@g…`.

**Comment by Hywan** [comment]
@jubianchi Maybe a rebase would be better (to avoid “CS” commits)?

**Pull request merged**